Contexts

ysimonson opened the pull request:
This introduces a new, backwards-incompatible version of configs, a migration to update old configs, and the related behavioral changes to `pachctl` from the new config.

Closes #3774 
Fixes #3538
Closes #3036
Fixes #3419

## Contexts

The largest addition is contexts, which are akin to `kubectl` contexts. Instead of building off of config V1 and having a different config file for each context (as proposed in #3774), this stores all contexts in a single config file. The reasons for this:
1) The hope with multiple files was that we could save effort by just building off of config V1. I no longer think that is the case, because bolting contexts on top of that design appears to be just as much work, if not more.
2) This design avoids ambiguity when a user sets `PACH_CONFIG` that the multiple config file design has.
3) It more closely follows k8s' approach.

Contexts have the same fields as config V1, plus a context source field, which specifies where contexts came from.

## Active context

Config V2 contains a reference to the currently active context. The active context can be overridden via the env var `PACH_CONTEXT`.

## Metrics

Rather than having a global flag `--no-metrics` that needs to be passed in for each call to `pachctl`, the ability to disable metrics are now specified in the config.

## Config implementation changes

Configs are now read only once per run of `pachctl`. Before, it was read multiple times, which allowed for subtle bugs (e.g. if the user ID wasn't yet set, it would be reset multiple times.) I did not add any further locking to ensure changes can't overwrite each other. Given the current uses of configs (which are predominantly read-only), I think this is safe _enough_ for now, but it is certainly not bulletproof!

## Migrations

A migration is run the first time the config is read. The V1 config is turned into a context.

## Deployments

When a (non-dry-run) deployment succeeds, a new pach context is created, and the user is automatically switched to the new context.

## New commands

* `pachctl config get metrics` - gets whether metrics are enabled
* `pachctl config set metrics (true|false)` - sets whether metrics are enabled
* `pachctl config get active-context` - gets the active context
* `pachctl config set active-context [name]` - sets the active context
* `pachctl config get context [name]` - gets a context config JSON by name
* `pachctl config set context [name] [--overwrite]` - sets a context config from JSON stdin
* `pachctl config update context [name] --pachd-address=[address]` - updates the pachd address of an existing context (this is the only field that is updatable without completely overwriting a context, at the moment.)
* `pachctl config delete context [name]` - removes a context
* `pachctl config list context` - lists all contexts

## Removals

The global `--no-metrics` and `--no-port-forwarding` flags were removed, in favor of a config values.

JoeyZwicker commented:
Are we definitely deciding to deprecate PACHD_ADDRESS? Or are we keeping it, but not having be the primary connecting option in the docs, etc?

ysimonson commented:
With the current approach, `PACHD_ADDRESS` will work, but warnings will be printed about its impending removal

Tryneus commented:
LGTM so far.  I'm slightly concerned that `PACHD_ADDRESS` might be important to certain workflows, and I'm generally in favor of being able to one-off-overide behavior on the command line through env vars.  However, I'm pretty ok with this for now - if someone really needs this we'll probably hear about it in the period of time that it's spitting out deprecation warnings.

ysimonson commented:
@Tryneus contexts include a pachd address, as well as auth credentials. The concern is that setting `PACHD_ADDRESS` makes it trivial to send auth credentials to the wrong cluster. This issue is somewhat (though not fully) mitigated with the approach taken here, where `PACHD_ADDRESS` is only respected on the context that was migrated from config V1.

ysimonson commented:
@Tryneus I need to add CLI endpoints for managing contexts, what do you think of the following?

pachctl get metrics - gets whether metrics are enabled
pachctl set metrics <value> - sets whether metrics are enabled
pachctl use context <context name> - sets the active context
pachctl get context (<name>|--active) - gets a context config by name, or gets the currently active context
pachctl put context <name> <config> - sets a context config
pachctl delete context <name> - removes a context
pachctl list context - lists all contexts

A few notes:
- Is get/set metrics too ambiguous? Will too many users think it actually gets/sets metrics data?
- It strays from k8s' CLI for managing contexts/configs, but I think less ambiguous and more aligned with how our CLI is setup with your changes.
- With this, contexts would be immutable (you can't change individual values in a context) but you could override a context.

Tryneus commented:
>pachctl get metrics - gets whether metrics are enabled
pachctl set metrics [value] - sets whether metrics are enabled

I'd prefer if these were more like:
* `pachctl metrics` - prints the current state of metrics
* `pachctl metrics --enable` - enables metrics
* `pachctl metrics --disable` - disables metrics

Reasoning:
 * less overloading of the `get` verb, which currently dumps data from the cluster, while this is a local-only configuration change
 * `set metrics [value]` feels a bit overly verbose

Similarly, I'm a little wary of managing contexts using the same verbs for managing resources in the pachyderm cluster, I think it will break expectations.  What if it was more like:

* `pachctl context` - prints the currently active context
* `pachctl context <name>` - activates a context by name
* `pachctl context --list [options]` - lists all local contexts
* `pachctl context --get <name> [options]` - prints detailed information about a context
* `pachctl context --add <name> [options]` - adds a context
* `pachctl context --remove <name> [options]` - removes a context

The biggest con I can see is that this breaks our current pattern of `<verb> <noun>`, but I _think_ that is alright because contexts are not Pachyderm-level resources, but abstractions at the CLI level.

ysimonson commented:
That makes sense. I find a couple of those commands ambiguous (most particularly the context activation one, as without a description I would've thought that it would get details of the given context), but can play around with the general idea and see.

jdoliner commented on src/client/pkg/config/config.go:
```diff
@@ -42,24 +60,67 @@ func Read() (*Config, error) {
 // file.
 func (c *Config) Write() error {
+	if c.V1 != nil {
+		panic("v1 config included, implying a bug")
```
Very minor nit, we generally use the string `(this is likely a bug)` in cases like this. I figure we might as well keep it consistent throughout.

jdoliner commented:
@Tryneus it seems kind of weird to me to have verbs as flags in metrics and context commands rather than as positional args like the rest of the commands, even though I do see the argument for why they're different compared to commands that interact with the cluster. Maybe it would make sense to put all of these under a subcommand called `config` (that's what `kubectl` uses for their equivalent, and while I don't think their config support is a paragon of good design I think the name is right.) That would save polluting the global namespace, which would maybe make it more tolerable to use pos-args instead of flags?

Tryneus commented:
> put all of these under a subcommand called `config`
@jdoliner - this solves my concerns and I am ok with it

ysimonson commented:
Alright, I'll move forward with the first send of commands, under the `config` subcommand

ysimonson commented:
One open question is the interaction points between this and kubectl contexts (if any.) Originally I was thinking that these could be totally isolated from kubectl contexts (i.e., adding a pach context does not add a kubectl context.) I don't think that's feasible because:
1) It would significantly hamper current workflows because users frequently use kubectl to introspect/modify the underlying k8s cluster. 
2) Pach's port forwarder would have to be rewritten, reinventing a lot of stuff that k8s' client library already does.

Given that, I'm now thinking that each context should contain a reference to its underlying kubectl context, and the creation or deletion of a pach context will reflect in the k8s context.

Are there any alternate designs that might make more sense?

ysimonson commented:
I've rewritten the PR description to be more in line with progress to date. There's still a few open design questions; I have bolded them in the description. Feedback welcome!

ysimonson commented:
@jdoliner Ready for review. No rush though, since this shouldn't be merged in until after 1.9 is out, and I want to kick the tires with it a bit anyways.

There's still one open design question - see the migrations section in the PR description.

ysimonson commented on src/server/cmd/pachctl/cmd/cmd.go:
```diff
@@ -404,9 +400,11 @@ Environment variables:
+				// TODO(ys): metrics used to be force-disabled here, do we still want that?
+				pachClient, err = client.NewOnUserMachine("user", client.WithDialTimeout(timeout))
 			} else {
-				pachClient, err = client.NewOnUserMachine(false, !noPortForwarding, "user")
```
@jdoliner metrics were disabled when creating a client for `pachctl version`. I'm not sure if this was a mistake, or by design. I'm assuming the former here, is that incorrect?

jdoliner replied:
Yeah, seems like a mistake to me.

ysimonson commented:
Is there a reason we have a `--namespace` argument when it could be set in the underlying k8s context? I'm tempted to remove it, since with these changes, we now rely on the context's namespace everywhere else.

ysimonson commented on src/server/config/cmds.go:
```diff
@@ -0,0 +1,305 @@
+	commands = append(commands, cmdutil.CreateAlias(listContext, "config list context"))
+
+	// TODO(ys): help template and usage func are nearly a copy-paste from
+	// what's in cmdutil/cobra.go, clean this up
```
@Tryneus this is kind of gnarly; it's a copy-paste of what's cmdutil/cobra.go, and the way the walk function finds associated commands is super hackey. If you've got ideas for how to DRY this up or otherwise clean it up, let me know.

ysimonson commented:
I'm debating whether `PACHD_ADDRESS` should be deprecated. I think the present behavior is a bit confusing (i.e., it only works if you're on a migrated profile.) @JoeyZwicker thoughts?

ysimonson commented:
I'm no longer deprecating `PACHD_ADDRESS`. We can always deprecate it later if need be.

Tryneus commented:
@ysimonson - this appears to be because `cobra.Command` inherits its parent's `UsageFunc` in the command hierarchy, which hasn't been a problem before because docs commands did not have children.  I'll prepare a PR fix.

ysimonson commented:
I've just removed the link between pach contexts and kube contexts. This was largely unused, and a discussion with Kevin made me realize that linking the two is problematic because there's a variety of ways in which the linked kube context might not be available at the time a `pachctl` command is issued (for example, if `KUBE_CONFIG` is set.)

In the process, I also cleaned things up related to namespaces, as outlined in the removal section of the PR description.

One issue I'm currently facing is that deployments aren't working because the call to `kubectl apply` in deployment-related commands doesn't seem to work unless a namespace is explicitly set. Would anyone know why that is?
